### ISSUE TITLE
Bottleneck with `IterableDataset` on many workers

### ISSUE THREAD

**Issue by lvwerra** [huggingface/accelerate#158]
### Issue
When preprocessing in the `IterableDataset` is computationally expensive it can become a bottleneck for training. The issue becomes more severe the more workers there are since the CPU resources are shared among the n workers and all workers preprocess all samples and then slice out. As implemented in the `IterableDatasetShard` that means that the preprocessing is done n times for all samples used in every worker. 

### Example
Especially when training small GPT models and tokenizing on the fly this becomes a bottleneck when training in a distributed setting. E.g. if I have a `bs_per_device=2` on `n_workers=16` the 32 samples of the global batch are tokenized 16 times. Keeping the `bs_per_device ` constant this scales quadratically with added workers. For large corpora pretokenizing becomes impractical and also wouldn't allow for data streaming.

### Solutions
- Is there an option in accelerate or PyTorch to a have a centralized dataloader to avoid redundant processing steps and make full use of the infrastructure?
- Could we spin up an independant script as kind of a data server that handles the `IterableDataset` and then have each worker requesting from this central service? Since it is decoupled from the GPU workers it would also have access to all CPU resources and because it is central it would only need to do the preprocessing once.

_Note:_ When using the streaming feature of datasets it can happen that due to a timeout to the hub an element is skipped. Because that not necessarily happens for all workers the dataloaders become out of sync and there could be duplicate samples across the workers. This issue would also be solved with a central dataloader.

**Comment by sgugger** [comment]
That is a limitation of an `IterableDataset`, which is why I never use that class on my side. There is no going around the fact that you need to preprocess sample 15 to then get access to sample 16.

The solution you propose would require a lot of code at first glance, and probably be its own library. There is no easy way to have a "central" object with torch distributed since the script is launched as many times as there are processes. The easiest solution is to just forego the streaming feature IMO and to rely on a regular dataset, for which processing is much more efficient. Also since tokenizing seems to be taking a lot of time, pre-tokenizing everything once for all seems like a good idea.

**Comment by sgugger** [comment]
Another solution would be to have the datasets library handle the sharding for you in the IterableDataset and then not passing your dataloader to the `accelerator.prepare` method so that `Accelerate` doesn't re-shard, but that's another feature to implement in Datasets just for the benefit of streaming, which seems like a bit of an overkill since you will need the whole dataset anyway.

**Comment by lvwerra** [comment]
One thing that might also help but is not directly related to `IterableDataset` would be an asynchronous dataloader such that while the model operations are performed the next batch is prepared instead of doing so sequentially. This would reduce the iteration time from `T = t_dataloader + t_model_ops` to `T = max(t_dataloader, t_model_ops)` but I have no idea how hard that would be to implement.

**Comment by lvwerra** [comment]
Regarding your comment concerning not using `IterableDataset` and streaming: the full dataset i am working with is around 200 GB big so while pretokenizing is possible it comes with challenges and streaming would be an elegant solution to that.

**Comment by sgugger** [comment]
Thinking of this a bit more, I could add an option to have the DataLoader be only created and iterred through on process 0, then have the final batches it creates sent on each processes. This shouldn't be too much work and respects your idea of a centralized dataloader, I believe. It will however add one sync at each step (to send to each process their part of the batch) on top of the all_reduce of the gradients, is that acceptable?

**Comment by lvwerra** [comment]
That sounds like a great idea! I don't know how much cost the extra sync step adds, but I guess that is a tradeoff between the two approaches that depends a bit on the setup. Having it as an option one could then explore which one works best depending on the infrastructure and preprocessing steps.

**Comment by sgugger** [comment]
I have put a proof of concept in this [branch](https://github.com/huggingface/accelerate/tree/main_dataloader). Could you give it a try and see if it solves your issue/works ok? The only change is to instantiate your `accelerator` with
```
accelerator = Accelerator(main_dataloader=True)
```

**Comment by ssharpe42** [comment]
Hey all, has there been any more research into how to make this more efficient?
